src/test/scala/chiselTests/DPISpec.scala

```diff
@@ -0,0 +1,110 @@
+package chiselTests.simulator
+
+import chisel3._
+import chisel3.experimental.ExtModule
+import chisel3.simulator._
+import chisel3.util.circt.dpi._
+import circt.stage.ChiselStage
+import chisel3.util.{HasExtModuleInline, HasExtModulePath, HasExtModuleResource}
+import org.scalatest.funspec.AnyFunSpec
+import org.scalatest.matchers.must.Matchers
+import org.scalatest.matchers.should.Matchers.convertToAnyShouldWrapper
+import chisel3.util.HasBlackBoxInline
+import svsim._
+
+class DPITest extends Module {
+  val dpiImpl = s"""
+                   |#include <stdint.h>
+                   |#include <iostream>
+                   |
+                   |extern "C" void hello()
+                   |{
+                   |    std::cout << "hello from c++\\n";
+                   |}
+                   |
+                   |extern "C" void add(int lhs, int rhs, int* result)
+                   |{
+                   |    *result = lhs + rhs;
+                   |}
+                   |
+                   |struct CounterState
+                   |{
+                   |    int counter = 0;
+                   |};
+                   |
+                   |extern "C" void create_counter(CounterState** result)
+                   |{
+                   |    std::cout << "Create counter\\n";
+                   |    *result = new CounterState;
+                   |}
+                   |
+                   |extern "C" void increment_counter(CounterState* state, int* result)
+                   |{
+                   |    *result = state->counter++;
+                   |}
+  """.stripMargin
+
+  val io = IO(new Bundle {
+    val a = Input(UInt(32.W))
+    val b = Input(UInt(32.W))
+    val add_result = Output(UInt(32.W))
+    val counter_result = Output(UInt(32.W))
+  })
+
+  // Void function
+  RawClockedVoidFunctionCall("hello")(clock, true.B)
+
+  // Stateless function with result
+  val result = RawClockedNonVoidFunctionCall("add", UInt(32.W))(clock, true.B, io.a, io.b)
+
+  // A function that allocates a state and returns a pointer.
+  val counter = RawClockedNonVoidFunctionCall("create_counter", UInt(64.W))(clock, reset.asBool)
```

Comment:
Originally in SystemVerilog, if put multiple `RawClockedNonVoidFunctionCall` in an always block, they will be executed in order. This is be used to schedule the order of different DPI calls with dependencies, e.g. peek `valid` signal and poke the `ready` signal for handshake. It seems that we need to merge two DPIs together to achieve this?

Reply:
> It seems that we need to merge two DPIs together to achieve this?

Exactly, that's recommended way to model DPI dependency. It's extremely hard and breaking change to represent dependency within always clock with the current FIRRTL representation.